sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -1966,21 +1966,33 @@ void ProgramManager::addImage(sycl_device_binary RawImg,
              "Invalid Bfloat16 Device Library Index.");
       if (m_Bfloat16DeviceLibImages[Bfloat16DeviceLibVersion].get())
         return;
-      size_t ImgSize =
-          static_cast<size_t>(RawImg->BinaryEnd - RawImg->BinaryStart);
-      std::unique_ptr<char[]> Data(new char[ImgSize]);
-      std::memcpy(Data.get(), RawImg->BinaryStart, ImgSize);
-      auto DynBfloat16DeviceLibImg =
-          std::make_unique<DynRTDeviceBinaryImage>(std::move(Data), ImgSize);
-      auto ESPropSet = getExportedSymbolPS(RawImg);
-      sycl_device_binary_property ESProp;
-      for (ESProp = ESPropSet->PropertiesBegin;
-           ESProp != ESPropSet->PropertiesEnd; ++ESProp) {
-        m_ExportedSymbolImages.insert(
-            {ESProp->Name, DynBfloat16DeviceLibImg.get()});
+
+      if (IsDeviceImageCompressed) {
+        // Decompress the image.
+        auto ESPropSet = getExportedSymbolPS(RawImg);
+        for (auto ESProp = ESPropSet->PropertiesBegin;
+             ESProp != ESPropSet->PropertiesEnd; ++ESProp) {
+          m_ExportedSymbolImages.insert({ESProp->Name, Img.get()});
+        }
+        CheckAndDecompressImage(Img.get());
```

Comment:
Does this need to happen after the exported symbols are inserted? If not, could we hoist the property handling loop and the `m_Bfloat16DeviceLibImages[Bfloat16DeviceLibVersion] = std::move(...` after the if-statement?

Reply:
That's a good point. Let me try hoisting the loop after the if-else statement.

Reply:
Done in https://github.com/intel/llvm/pull/18108/commits/a7b72eef10cdb1ec863e96c4262def2ad706a38f